web/server/web_client.c

```diff
@@ -812,109 +849,348 @@ static inline char *http_header_parse(struct web_client *w, char *s, int parse_u
+    }
+
+    return s;
+}
+
+/**
+ *  Parse header
+ *
+ *  Parse the headers sent together with the URL.
+ *
+ * @param w is the structure with all information of the client request.
+ * @param s is the string where the URL finished
+ */
+static inline void web_client_parse_headers(struct web_client *w,char *s) {
+    while(*s) {
+        // find a line feed
+        while(*s && *s++ != '\r');
+
+        // did we reach the end?
+        if(unlikely(!*s)) break;
+
+        if (*s == '\n') {
+            s++;
+        }
+
+        s = http_header_parse(w, s,
+                              (w->mode == WEB_CLIENT_MODE_STREAM) // parse user agent
+        );
+
+    }
+}
+
+/**
+ *  Parse request
+ *
+ * @param w is the structure with all information of the client request.
+ * @param divisor the place with the first equal sign
+ *
+ * @return It returns the number of variables in the query string.
+ */
+int web_client_parse_request(struct web_client *w,char *divisor) {
+    if(!divisor) {
+        w->total_params = 0;
+        return 0;
+    }
+
+    uint32_t i = url_parse_query_string(w->param_name,w->param_values,w->query_string.body+1,divisor);
+    w->total_params = i;
+
+    return i;
+}
+
+/**
+ *  Set Directory
+ *
+ *  Map directory variables
+ *
+ * @param w is the structure with all information of the client request.
+ * @param begin is the end of the first directory
+ * @param enddir a pointer to the position with the last valid character
+ * @param endcmd the end of the last character.
+ */
+static inline void web_client_set_directory(struct web_client *w,char *begin,char *enddir,char *endcmd) {
+    if (enddir) {
+        w->directory.body = begin;
+        w->directory.length = enddir - begin;
+
+        if (!strncmp(w->directory.body,"api",w->directory.length)) {
+            begin = enddir + 1;
+            enddir = strchr(begin,'/');
+            if(enddir) {
+                w->version.body = begin;
+                w->version.length = enddir - begin;
+
+                enddir++;
+                w->command.body = enddir;
+                w->command.length = (size_t) (endcmd - enddir);
+            }
+        }
+    }
+    else{
+        w->directory.body = begin;
+        w->directory.length = w->path.length - 1;
+        w->version.body = NULL;
+        w->version.length = 0;
+        w->command.body = NULL;
+        w->command.length = 0;
+    }
+}
+
+/**
+ *  Set without query string
+ *
+ *  When I do not have query string, it is necessary to set some default values.
+ *
+ * @param w is the structure with all information of the client request.
+ */
+static inline void web_client_set_without_query_string(struct web_client *w) {
+    w->query_string.body = NULL;
+    w->query_string.length = 0;
+
+    char *test = w->path.body+1;
+    if (!strncmp(test,"api/v1/",7) ) {
+        test += 7;
+        char *endofcommand=strchr(test,' ');
+        w->command.length=(endofcommand)?(size_t)endofcommand-(size_t)test:strlen(test);
```

Comment:
Logic: you want to do pointer arithmetic here so `(size_t)endofcommand-(size_t)test` should be `(size_t)(endofcommand - test)`.

Reply:
Both cases will give the correct result, but I agree that it is more readable this way (Am I learning correctly the concept of readable?)

Reply:
This can be worse cause some compilers givevsize_t be 32 bit when the pointers are 64 bit and if you're unlucky you get bogus numbers. The correct thing is to do pointer arithmetic before an integer cast.